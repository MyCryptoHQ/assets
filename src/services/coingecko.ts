import { Asset } from '../constants';
import { CachedResponse, getInitialCache, withCache } from './cache';
import { isSimilar } from '../utils';

const COINGECKO_API_ENDPOINT = 'https://api.coingecko.com/api/v3/coins/list';

interface CoinGeckoCoin {
  id: string;
  symbol: string;
  name: string;
}

interface CoinGeckoData {
  coins: CoinGeckoCoin[];
}

/**
 * Match an asset with a CoinGecko ID.
 *
 * @param {Partial<CoinGeckoData>} cache
 * @param {Asset} asset
 * @return {Promise<CachedResponse<CoinGeckoData, string | null>>}
 */
export const matchCoinGeckoId = async (
  cache: Partial<CoinGeckoData>,
  asset: Asset
): Promise<CachedResponse<CoinGeckoData, string | null>> => {
  const data = cache;
  if (!data || !data.coins) {
    data.coins = await getInitialCache<CoinGeckoCoin[]>(COINGECKO_API_ENDPOINT);
  }

  const match = data.coins.find(coin => coin.symbol === asset.symbol.toLowerCase());
  if (match && isSimilar(match.name, asset.name)) {
    return {
      cache: data,
      data: match.id
    };
  }

  return {
    cache: data,
    data: null
  };
};

export default withCache<CoinGeckoData, typeof matchCoinGeckoId, string | null>(matchCoinGeckoId);
