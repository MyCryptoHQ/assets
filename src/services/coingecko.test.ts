import { matchCoinGeckoId } from './coingecko';
import { matchCoinGeckoId as matchCoinGeckoIdWithCache } from './index';
import { getInitialCache } from './cache';

const CACHE_DATA = [{ id: 'ethereum', symbol: 'eth', name: 'Ethereum' }];

jest.mock('./cache', () => ({
  getInitialCache: jest.fn(async () => CACHE_DATA),

  withCache: jest.requireActual('./cache').withCache
}));

afterEach(() => {
  jest.clearAllMocks();
});

describe('matchCoinGeckoId', () => {
  it('initializes the cache', async () => {
    const asset = {
      name: 'Ethereum',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    };

    const { cache } = await matchCoinGeckoId({}, asset);
    const result = await matchCoinGeckoId(cache, asset);

    expect(result.cache.coins).toMatchSnapshot();
    expect(getInitialCache).toHaveBeenCalledTimes(1);
  });

  it('fetches the ID of an asset', async () => {
    const asset = {
      name: 'Ethereum',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    };

    await expect(matchCoinGeckoId({}, asset)).resolves.toEqual({
      cache: {
        coins: CACHE_DATA
      },
      data: 'ethereum'
    });
  });

  it('returns null if an asset could not be found', async () => {
    const asset = {
      name: 'Ethereum Classic',
      symbol: 'ETC',
      uuid: '6c1e671f-9af5-546d-9c1a-52067bdf0e53'
    };

    await expect(matchCoinGeckoId({}, asset)).resolves.toEqual({
      cache: {
        coins: CACHE_DATA
      },
      data: null
    });
  });

  it('works with a cache', async () => {
    const asset = {
      name: 'Ethereum',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    };

    await expect(matchCoinGeckoIdWithCache(asset)).resolves.toBe('ethereum');
  });
});
