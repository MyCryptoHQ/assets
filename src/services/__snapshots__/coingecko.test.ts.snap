// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchCoinGeckoId initializes the cache 1`] = `
Array [
  Object {
    "id": "ethereum",
    "name": "Ethereum",
    "symbol": "eth",
  },
]
`;
